src/ast/KnownGlobal.zig

```diff
@@ -8,26 +8,144 @@ pub const KnownGlobal = enum {
+
+                if (n == 1) {
+                    const arg = e.args.ptr[0];
+                    switch (arg.data) {
+                        .e_object, .e_array => {
+                            // new Object({a: 1}) -> {a: 1}
+                            // new Object([1, 2]) -> [1, 2]
+                            return arg;
+                        },
+                        .e_null, .e_undefined => {
+                            // new Object(null) -> {}
+                            // new Object(undefined) -> {}
+                            return js_ast.Expr.init(E.Object, E.Object{}, loc);
+                        },
+                        else => {},
+                    }
+                }
+
+                // For other cases, just remove 'new'
+                return callFromNew(e, loc);
+            },
+
+            .Array => {
+                const n = e.args.len;
+
+                if (n == 0) {
+                    // new Array() -> []
+                    return js_ast.Expr.init(E.Array, E.Array{}, loc);
+                }
+
+                // new Array(1, 2, 3) -> [1, 2, 3]
+                // But NOT new Array(3) which creates an array with 3 empty slots
+                if (n > 1) {
+                    return js_ast.Expr.init(E.Array, .{ .items = e.args }, loc);
+                }
+
+                // For single argument, only convert to literal if we're SURE it's not a number
+                if (n == 1) {
+                    const arg = e.args.ptr[0];
+
+                    // Check if it's an object or array literal first
+                    switch (arg.data) {
+                        .e_object, .e_array => {
+                            // new Array({}) -> [{}], new Array([1]) -> [[1]]
+                            // These are definitely not numbers, safe to convert
+                            return js_ast.Expr.init(E.Array, .{ .items = e.args }, loc);
+                        },
+                        else => {},
+                    }
+
+                    // For other types, check via knownPrimitive
+                    const primitive = arg.knownPrimitive();
+                    // Only convert if we know for certain it's not a number
+                    // unknown could be a number at runtime, so we must preserve Array() call
+                    switch (primitive) {
+                        .null, .undefined, .boolean, .string, .bigint => {
+                            // These are definitely not numbers, safe to convert
+                            return js_ast.Expr.init(E.Array, .{ .items = e.args }, loc);
+                        },
+                        .number, .unknown, .mixed => {
+                            // Could be a number, preserve Array() call
+                            return callFromNew(e, loc);
+                        },
+                    }
+                }
+
+                // For new Array(number), just remove 'new'
+                return callFromNew(e, loc);
```

Comment:
this is unreachable. i think this should be `switch (n) { 0 => {...}, 1 => {...}, else => {...}}`